### PR TITLE
Address mcapp RBAC changes

### PR DESCRIPTION
**Pull request by gitlawr**
Address issue:
https://github.com/rancher/rancher/issues/17147
https://github.com/rancher/rancher/issues/17981

Related types PR: https://github.com/rancher/types/pull/726

- Add subcommand for updating target projects
- Add subcommand for updating members
- Add --member, --member-access-type option on install
- Add --role option on install/upgrade
- Add member/role details in show-app command

Examples:
```
#update targets command:
1. rancher mcapp add-project myapp mycluster:Project1 mycluster:Project2
2. rancher mcapp delete-project myapp mycluster:Project1 mycluster:Project2
# update member command:
3. rancher mcapp add-member myapp owner user1 user2
4. rancher mcapp delete-member myapp user1 user2
# set members on installation
5. rancher mcapp install --member user1 --member-access-type member --target mycluster:Default mysql myapp
# set roles on install/upgrade
6. rancher mcapp install --role project-owner --target mycluster:Default mysql myapp
7. rancher mcapp upgrade --role project-owner myapp 0.3.8
# show app command
8. rancher mcapp sa myapp
CURRENT   REVISION              CREATED
*         mcapprevision-rhcw9   18 Feb 2019 05:10:39 UTC
          mcapprevision-gmzgh   18 Feb 2019 05:10:00 UTC

CURRENT   VERSION
          0.3.7
*         0.3.8

MEMBER_NAME     MEMBER_TYPE   ACCESS_TYPE
Default Admin   Local User    owner

ROLE_NAME
cluster-owner
project-owner
```

### CODE REVIEW

File: cmd/multiclusterapp.go [hunk 1]
```diff
@@ -101,6 +112,19 @@ func MultiClusterAppCommand() cli.Command {
...
+					cli.StringSliceFlag{
+						Name:  "role",
+						Usage: "Set roles required to view the app",
+					},
+					cli.StringSliceFlag{
+						Name:  "member",
+						Usage: "Set members who can view the app, with access type defined by --member-access-type",
+					},
+					cli.StringFlag{
+						Name:  "member-access-type",
```
Comment (orangedeng) [inline cmd/multiclusterapp.go]: Should each members have their own access type?

File: cmd/multiclusterapp.go [hunk 2]
```diff
@@ -138,17 +162,59 @@ func MultiClusterAppCommand() cli.Command {
...
 			},
+			cli.Command{
+				Name:        "delete-project",
+				Usage:       "Delete target projects from a multi-cluster app",
+				Action:      deleteMcappTargetProject,
+				Description: "Examples:\n #Delete 'p1' project in cluster 'mycluster' from target projects of a multi-cluster app named 'myapp'\n rancher multiclusterapp delete-project myapp mycluster:p1\n",
+				ArgsUsage:   "[APP_NAME/APP_ID, CLUSTER_NAME:PROJECT_NAME/PROJECT_ID...]",
+			},
+			cli.Command{
+				Name:        "add-member",
```
Comment (orangedeng) [inline cmd/multiclusterapp.go]: Should add-member also have the member-access-type parameter?

File: cmd/multiclusterapp.go (hunk 1 above)
Comment (gitlawr) [reply cmd/multiclusterapp.go]: It's not easy to claim different members with different access type in a single command. 
Therefore,
1. These arguments are for setting a set of members with the same access type on install.
2. We introduce add/delete-member to do accurate membership assignments

File: cmd/multiclusterapp.go (hunk 2 above)
Comment (gitlawr) [reply cmd/multiclusterapp.go]: Access-type is included in the command args. The usage of add-member is:
```
"[APP_NAME/APP_ID, USER_NAME/USER_ID, ACCESS_TYPE]"
rancher mcapp add-member user1 read-only
```
Or do you think it's better if it goes as
```
rancher mcapp add-member --member-access-type read-only u1 u2 u3...
```

Comment (orangedeng) [reply cmd/multiclusterapp.go]: If so, I would prefer using rancher mcapp add-member myapp owner user1 user2 ... Put the scaleable parameters in the end then you can add multiple user in this command.

Comment (gitlawr) [reply cmd/multiclusterapp.go]: Updated

**Comment by orangedeng** [comment]
LGTM

**Comment by gitlawr** [comment]
Types RP is merged. Rebased.

**Comment by cjellick** [comment]
@dramich to review

File: cmd/multiclusterapp.go [hunk 3]
```diff
@@ -679,45 +763,203 @@ func fromMultiClusterAppAnswers(answers []managementClient.Answer) map[string]st
...
+	app, input, err := proceedUpdateMultiClusterAppTargetsArgs(ctx, c)
+	if err != nil {
+		return err
+	}
+	return c.ManagementClient.MultiClusterApp.ActionRemoveProjects(app, input)
+}
+
+func proceedUpdateMultiClusterAppTargetsArgs(ctx *cli.Context, c *cliclient.MasterClient) (*managementClient.MultiClusterApp, *managementClient.UpdateMultiClusterAppTargetsInput, error) {
+	if len(ctx.Args()) < 2 {
+		return nil, nil, cli.ShowSubcommandHelp(ctx)
```
Comment (dramich) [inline cmd/multiclusterapp.go]: `cli.ShowSubcommandHelp(ctx)` is not an error, it prints the help for the command, so the error checks in `addMcappTargetProject` and `deleteMcappTargetProject` don't catch it which causes a panic further down. Check the length of args in the first function.

File: cmd/multiclusterapp.go [hunk 4]
```diff
@@ -679,45 +763,203 @@ func fromMultiClusterAppAnswers(answers []managementClient.Answer) map[string]st
...
+	}
+
+	for _, name := range memberNames {
+		member, err := searchForMember(ctx, c, name)
+		if err != nil {
+			return err
+		}
+
+		memberID := member.ID
+		toKeepMembers := make([]managementClient.Member, 0, len(app.Members))
```
Comment (dramich) [inline cmd/multiclusterapp.go]: This isn't needed. `var toKeepMembers []managementClient.Member` is fine for this.

File: cmd/multiclusterapp.go [hunk 5]
```diff
@@ -803,6 +1030,46 @@ func outputMultiClusterAppRevisions(ctx *cli.Context, c *cliclient.MasterClient)
 	return writer.Err()
 }
 
+func outputMultiClusterAppMembers(ctx *cli.Context, c *cliclient.MasterClient, app *managementClient.MultiClusterApp) error {
```
Comment (dramich) [inline cmd/multiclusterapp.go]: I think this should be its own command `list-members` following what is done in project. The `show-app` is about version info and if there are a lot of users on that app it's going to make for a large output.

File: cmd/multiclusterapp.go [hunk 6]
```diff
@@ -803,6 +1030,46 @@ func outputMultiClusterAppRevisions(ctx *cli.Context, c *cliclient.MasterClient)
...
+		writer.Write(&MemberData{
+			Name:       principal.Name,
+			MemberType: strings.Title(fmt.Sprintf("%s %s", principal.Provider, principal.PrincipalType)),
+			AccessType: m.AccessType,
+		})
+	}
+	return writer.Err()
+}
+
+func outputMultiClusterAppRoles(ctx *cli.Context, c *cliclient.MasterClient, app *managementClient.MultiClusterApp) error {
```
Comment (dramich) [inline cmd/multiclusterapp.go]: Hide this behind a flag on `show-app`

File: cmd/multiclusterapp.go (hunk 2 above)
Comment (dramich) [reply cmd/multiclusterapp.go]: Agreed on keeping it this way, it's how project and cluster members work.

File: cmd/multiclusterapp.go [hunk 7]
```diff
@@ -101,6 +112,19 @@ func MultiClusterAppCommand() cli.Command {
 						Name:  "target,t",
 						Usage: "Target project names/ids to install the app into",
 					},
+					cli.StringSliceFlag{
+						Name:  "role",
+						Usage: "Set roles required to view the app",
```
[the remark this answers is not in the document]
Comment (gitlawr) [reply cmd/multiclusterapp.go]: This is the same as what we display in the UI. Probably want to hide implementation details of the sa? @mrajashree Can you confirm if it's good or we need to tune the UI as well

[the remark this answers is not in the document]
Comment (mrajashree) [reply cmd/multiclusterapp.go]: Yes the description seems incorrect. These roles are what the system account or creator of the actual app instances will get. So maybe we can change it to "Set roles required to launch/manage the app"?

File: cmd/multiclusterapp.go (hunk 5 above)
Comment (gitlawr) [reply cmd/multiclusterapp.go]: Do you expect a `list-roles` command to display roles info of a mcapp

Comment (gitlawr) [reply cmd/multiclusterapp.go]: nevermind, just saw your comment below

**Comment by gitlawr** [comment]
@dramich Addressed above comments.

File: cmd/multiclusterapp.go [hunk 8]
```diff
@@ -138,17 +162,70 @@ func MultiClusterAppCommand() cli.Command {
...
+				Usage:       "Delete members from a multi-cluster app",
+				Action:      deleteMcappMember,
+				Description: "Examples:\n #Delete the membership of a user named 'user1' from a multi-cluster app named 'myapp'\n rancher multiclusterapp delete-member myapp user1\n",
+				ArgsUsage:   "[APP_NAME/APP_ID, USER_NAME/USER_ID...]",
+			},
+			cli.Command{
+				Name:        "list-members",
+				Aliases:     []string{"lm"},
+				Usage:       "List current members of a multi-cluster app",
+				Description: "List current members of a multi-cluster app",
```
Comment (dramich) [inline cmd/multiclusterapp.go]: Either update description to be different or drop it, having them the same makes the help output look funny.

File: cmd/multiclusterapp.go [hunk 9]
```diff
@@ -679,6 +778,156 @@ func fromMultiClusterAppAnswers(answers []managementClient.Answer) map[string]st
...
+	return input, nil
+}
+
+func addMcappMember(ctx *cli.Context) error {
+	if len(ctx.Args()) < 3 {
+		return cli.ShowSubcommandHelp(ctx)
+	}
+
+	appName := ctx.Args().First()
+	accessType := ctx.Args().Get(1)
```
Comment (dramich) [inline cmd/multiclusterapp.go]: Update this to use `strings.ToLower`

File: cmd/multiclusterapp.go [hunk 10]
```diff
@@ -522,6 +594,33 @@ func multiClusterAppTemplateInstall(ctx *cli.Context) error {
 	}
 	app.TemplateVersionID = templateVersionID
 
+	roles := ctx.StringSlice("role")
+	if len(roles) > 0 {
+		app.Roles = roles
+	}
+
+	members := ctx.StringSlice("member")
+	accessType := ctx.String("member-access-type")
```
Comment (dramich) [inline cmd/multiclusterapp.go]: Update this to use `strings.ToLower` to make the comparison nicer for the user.

**Comment by gitlawr** [comment]
@dramich Updated

**Comment by dramich** [comment]
LGTM

File: cmd/multiclusterapp.go [hunk 11]
```diff
@@ -34,12 +36,21 @@ Example:
...
+
 type MultiClusterAppData struct {
+	ID      string
 	App     managementClient.MultiClusterApp
 	Version string
 	Targets string
 }
 
+type MemberData struct {
+	Name       string
```
Comment (mrajashree) [inline cmd/multiclusterapp.go]: Member has different fields for groupPrincipalId and userPrincipalId https://github.com/rancher/types/blob/master/apis/management.cattle.io/v3/multi_cluster_app.go#L56
How do we differentiate between them here?

File: cmd/multiclusterapp.go [hunk 12]
```diff
@@ -138,17 +162,69 @@ func MultiClusterAppCommand() cli.Command {
 						Name:  "set",
 						Usage: "Set answers for the template, can be used multiple times. Example: --set foo=bar",
 					},
+					cli.StringSliceFlag{
+						Name: "role,r",
+						Usage: "Set roles required to launch/manage the apps in target projects. Specified roles on " +
+							"upgrade will override all the original roles. Leave it empty to keep current roles",
```
Comment (mrajashree) [inline cmd/multiclusterapp.go]: Why are roles on upgrade overriding all existing roles? If user has to add more roles will they be expected to provide all existing roles as input again? If so then can we add that as well to the usage?

File: cmd/multiclusterapp.go (hunk 11 above)
Comment (gitlawr) [reply cmd/multiclusterapp.go]: It is shown in `memberType`, See L1002:
```
MemberType: strings.Title(fmt.Sprintf("%s %s", principal.Provider, principal.PrincipalType))
```

File: cmd/multiclusterapp.go (hunk 12 above)
Comment (gitlawr) [reply cmd/multiclusterapp.go]: It's a simpler way to update roles. Or we will need to provide different flags/commands to add or delete roles of a mcapp. Updated the usage.

**Comment by gitlawr** [comment]
Squashed

**Comment by mrajashree** [comment]
LGTM

**Pull request merged**